Cabal-syntax/src/Distribution/Types/ComponentName.hs

```diff
@@ -20,11 +21,11 @@ import qualified Text.PrettyPrint as Disp
 import qualified Distribution.Compat.CharParsing as P
 
 -- Libraries live in a separate namespace, so must distinguish
-data ComponentName = CLibName   LibraryName
-                   | CFLibName  UnqualComponentName
-                   | CExeName   UnqualComponentName
-                   | CTestName  UnqualComponentName
-                   | CBenchName UnqualComponentName
+data ComponentName = CLibName   { toLibName :: LibraryName }
```

Comment:
I’m worried about introduction of partial field accessors. The proper solution would be to factor out the other four fields into a separate data type (`NonLibComponentName` or some such), which could have a total accessor. Didn’t check how much churn that would involve.

Reply:
If throw in some pattern synonyms on the top, the churn could actually be none.

Reply:
Perhaps we can forego the field accessors entirely, given a function like
```haskell
extract :: ComponentName -> Either LibraryName UnqualComponentName
extract (CLibName   s) = Left s
extract (CFLibName  s) = Right s
extract (CExeName   s) = Right s
extract (CTestName  s) = Right s
extract (CBenchName s) = Right s
```
that way we can avoid redefining types while keeping `componentNameRaw` and `componentNameString` simple.

Reply:
I like @ulysses4ever 's proposal.  One can define pattern synonyms for `CFLibName` etc. so that the change to the codebase remains small.

Reply:
Could either of you show me what you have in mind? Perhaps my lack of understanding but I can't see the change to the codebase being smaller.

Reply:
`Either`-based idea is not bad too, but here's an example of what I meant (can compile and run to get `"x"`)

```
{-# LANGUAGE PatternSynonyms #-}

type LibraryName = String
type UnqualComponentName = String

data NotLibComponentName
        = CNLFLibName  UnqualComponentName
        | CNLExeName   UnqualComponentName
        | CNLTestName  UnqualComponentName
        | CNLBenchName UnqualComponentName
data ComponentName
        = CLibName   LibraryName
        | CNotLibName NotLibComponentName
pattern CFLibName n = CNotLibName (CNLFLibName n)
// ... 3 more of these

main = print n
  where
    c = CFLibName "x" -- using bidirectional pattern to construct
    (CFLibName n) = c -- using bidirectional pattern to destruct
```

Reply:
(And then change `NotLibComponentName` to have a field accessor, and possibly some convenience function for `ComponentName`)